library/std/src/os/windows/io/handle.rs

```diff
@@ -0,0 +1,390 @@
+//! Owned and borrowed OS handles.
+
+#![unstable(feature = "io_safety", issue = "87074")]
+
+use super::raw::{AsRawHandle, FromRawHandle, IntoRawHandle, RawHandle};
+use crate::convert::TryFrom;
+use crate::ffi::c_void;
+use crate::fmt;
+use crate::fs;
+use crate::marker::PhantomData;
+use crate::mem::forget;
+use crate::ptr::NonNull;
+use crate::sys::c;
+use crate::sys_common::{AsInner, FromInner, IntoInner};
+
+/// A borrowed handle.
+///
+/// This has a lifetime parameter to tie it to the lifetime of something that
+/// owns the handle.
+///
+/// This uses `repr(transparent)` and has the representation of a host handle,
+/// so it can be used in FFI in places where a handle is passed as an argument,
+/// it is not captured or consumed, and it is never null.
+///
+/// Note that it *may* have the value `INVALID_HANDLE_VALUE`. See [here] for
+/// the full story.
+///
+/// [here]: https://devblogs.microsoft.com/oldnewthing/20040302-00/?p=40443
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+#[unstable(feature = "io_safety", issue = "87074")]
+pub struct BorrowedHandle<'handle> {
+    raw: NonNull<c_void>,
+    _phantom: PhantomData<&'handle OwnedHandle>,
+}
+
+/// An owned handle.
+///
+/// This closes the handle on drop.
+///
+/// This uses `repr(transparent)` and has the representation of a host handle,
+/// so it can be used in FFI in places where a handle is passed as a consumed
+/// argument or returned as an owned value, and is never null.
+///
+/// Note that it *may* have the value `INVALID_HANDLE_VALUE`. See [here] for
+/// the full story. For APIs like `CreateFileW` which report errors with
+/// `INVALID_HANDLE_VALUE` instead of null, use [`OptionFileHandle`] instead
+/// of `Option<OwnedHandle>`.
+///
+/// [here]: https://devblogs.microsoft.com/oldnewthing/20040302-00/?p=40443
+#[repr(transparent)]
+#[unstable(feature = "io_safety", issue = "87074")]
+pub struct OwnedHandle {
+    raw: NonNull<c_void>,
+}
+
+/// Similar to `Option<OwnedHandle>`, but intended for use in FFI interfaces
+/// where `INVALID_HANDLE_VALUE` is used as the sentry value, and null values
+/// are not used at all, such as in the return value of `CreateFileW`.
+///
+/// If this holds an owned handle, it closes the handle on drop.
+///
+/// This uses `repr(transparent)` and has the representation of a host handle,
+/// so it can be used in FFI in places where a non-null handle is passed as a
+/// consumed argument or returned as an owned value, or it is
+/// `INVALID_HANDLE_VALUE` indicating an error or an otherwise absent value.
+///
+#[repr(transparent)]
+#[unstable(feature = "io_safety", issue = "87074")]
+pub struct OptionFileHandle {
```

Comment:
I guess this wasn't explicitly mentioned, but it is effectively part of the Windows counterpart to `OwnedFd`, which we said we wanted. Because Windows has this surprising behavior where some APIs use `INVALID_HANDLE_VALUE` to indicate failure and other APIs use `INVALID_HANDLE_VALUE` as a valid handle value, it's useful to have two different types to be able to describe the two different kinds of APIs.

Reply:
Also see my answer above, but I think this is mostly a result from system call thing. With that I mean the caller needs to check if the handle is valid when it's created (i.e. after a system call), after that is can just use `OwnedHandle`.

Reply:
That's essentially how `OptionFileHandle` is meant to be used too. It's just for [use it in an FFI declaration](https://github.com/sunfishcode/io-lifetimes/blob/main/src/example_ffi.rs#L43) of functions that use `INVALID_FILE_HANDLE` for errors, and the main way to use it is to [immediately convert it to an `OwnedHandle`](https://github.com/sunfishcode/io-lifetimes/blob/main/examples/hello.rs#L79), which does the check for `INVALID_HANDLE_VALUE`. So the person writing the FFI declaration has use the right type, but then all call sites are prevented from doing the wrong check. Does that make sense? I'll add some comments explaining that in more detail.

Reply:
I don't these types should be used in ffi at all to be honest. Only once the caller (of the ffi function) checked all errors, validity, etc. should these types come in to play. Similar to how allocation functions use `*mut u8` and convert it into e.g. `Box<T>` after.

Reply:
I summarized this concern in a top-level comment [here](https://github.com/rust-lang/rust/pull/87329#issuecomment-887688834) to give it more visibility.